src/test/java/org/javarosa/xpath/analysis/test/TreeReferenceAccumulatorTest.java

```diff
@@ -0,0 +1,157 @@
+                XPathReference.getPathExpr("instance('baseinstance')/base/element").getReference());
+
+        EvaluationContext nestedRoot = new EvaluationContext(root,
+                XPathReference.getPathExpr("instance('nestedinstance')/nested/element").getReference());
+
+
+        runAndTest(root,
+                "instance('expr')/ebase/element[@attr = current()/context_value]/value",
+                "instance('expr')/ebase/element/value",
+                "instance('expr')/ebase/element/@attr",
+                "instance('baseinstance')/base/element/context_value");
+
+        runAndTest(nestedRoot,
+                "instance('expr')/ebase/element[@attr = current()/context_value]/value",
+                "instance('expr')/ebase/element/value",
+                "instance('expr')/ebase/element/@attr",
+                "instance('nestedinstance')/nested/element/context_value");
+
+        runAndTest(root,
+                "../other_element",
+                "instance('baseinstance')/base/other_element");
+
+
+        runAndTest(root,
+                "instance('cars')/cars/car[../@lead_make = instance('models')/models/model[@year = current()/year]/make]/name",
+                "instance('cars')/cars/@lead_make",
+                "instance('cars')/cars/car/name",
+                "instance('models')/models/model/@year",
+                "instance('models')/models/model/make",
+                "instance('baseinstance')/base/element/year"
+                );
+
+
+
+
+    }
+
+
+
+    @Test
+    public void testInvalidReferences() throws XPathSyntaxException {
+        runForError("../relativeRef");
+        runForError("current()/relative");
+    }
+
+
+
+
+    private void runForError(String text) {
+        Set<TreeReference> references = runAndTest(text);
+        if (references != null) {
+            Assert.fail(String.format("Should have failed to analyse expression %s", text));
+        }
+    }
+
+
+    private Set<TreeReference> runAndTest(String text, String... matches) {
+        return runAndTest(null, text, matches);
+    }
+
+    /**
+     * Tests that running TreeReferenceAccumulatingAnalyzer on the XPath expression represented
+     * by @text results in accumulating all of the tree refs listed in @matches
+     */
+    private Set<TreeReference> runAndTest(EvaluationContext context, String text, String... matches)  {
+        XPathExpression expression;
+        try {
+            expression = XPathParseTool.parseXPath(text);
+        } catch(XPathSyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        TreeReferenceAccumulatingAnalyzer analyzer;
+        if (context == null) {
+            analyzer = new TreeReferenceAccumulatingAnalyzer();
+        } else {
+            analyzer = new TreeReferenceAccumulatingAnalyzer(context);
+        }
+
+        Set<TreeReference> results = analyzer.accumulate(expression);
+        if (results == null) {
+            return null;
```

Comment:
Why does this just return null? If there is anything in `matches`, shouldn't the test fail at this point?

Reply:
hm, good point. fixing

Reply:
Cool. Could you also add a test that expects an analysis failure? similar to https://github.com/dimagi/commcare-core/blob/master/src/test/java/org/javarosa/xpath/analysis/test/StaticAnalysisTest.java#L86-L88

Reply:
The tests in testInvalidReferences() on line 98 test for analysis failure.